## Credit Card Fraud Detection


Build a machine learning model to identify fraudulent credit card
transactions.
Preprocess and normalize the transaction data, handle class
imbalance issues, and split the dataset into training and testing sets.
Train a classification algorithm, such as logistic regression or random
forests, to classify transactions as fraudulent or genuine.
Evaluate the model's performance using metrics like precision, recall,
and F1-score, and consider techniques like oversampling or
undersampling for improving results.

In [1]:
## importing 

In [51]:
!pip install scikit-learn
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression 
from sklearn.metrics import accuracy_score 
from sklearn.metrics import classification_report, confusion_matrix


[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [5]:
# loading the dataset into Dataframe 

In [4]:
credit_cart_data = pd.read_csv(r'D:\Data Analysist\Python\ML_In _Python Credit Card Fraud Detection\creditcard.csv')

In [5]:
credit_cart_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [6]:
credit_cart_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [7]:
credit_cart_data.shape

(284807, 31)

In [16]:
# analysis missing Data in the Data set 

In [8]:
credit_cart_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [9]:
credit_cart_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [18]:
# how many legal Transactions and fraduals Transaction Happned 

In [10]:
credit_cart_data['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

In [22]:
## this data set it very unbalanced so handling unblance 1 label 
##seperate all the narmal transaction and fraudulent transaction 


In [11]:
legit = credit_cart_data[credit_cart_data.Class == 0]
fraud = credit_cart_data[credit_cart_data.Class == 1]

In [12]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [26]:
# Statistical Measures

In [13]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [14]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [16]:
credit_cart_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

In [30]:
#handling unblance
#Under-Sampling Method used 
#Build a sample dataset containing similar distriution of normal transaction and Fraudulent Transactions(492)
#492 for normal and fraudlent transaction take for better prediction in ML

In [17]:
legit_sample = legit.sample(n=492)

In [32]:
# concatenating Two data set legit_sample and fraud

In [18]:
new_dataset = pd.concat([legit_sample , fraud], axis = 0)

In [19]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
7014      9195.0  1.174829 -0.027781  1.028653  1.056980 -0.552270  0.189836   
141758   84482.0  1.305574 -0.408376  0.548300 -0.917218 -0.847102 -0.357029   
236261  148737.0  1.811565 -0.665829 -0.466436  0.422015 -0.900105 -0.818862   
113513   73095.0  0.077719 -0.163961 -0.073190 -2.069075  2.015091  3.419668   
255101  157064.0  1.844602 -0.051082 -1.690806  1.465939  0.359141 -1.108321   

              V7        V8        V9  ...       V21       V22       V23  \
7014   -0.520784  0.030701  2.178297  ... -0.265712 -0.103326 -0.021588   
141758 -0.625507  0.049831  1.957069  ... -0.073461  0.017461 -0.133228   
236261 -0.397016 -0.048057  1.529336  ... -0.089888 -0.416063  0.280788   
113513 -0.267601  0.598043 -1.411262  ...  0.335748  0.932790 -0.239944   
255101  0.791471 -0.461617  0.144801  ...  0.111221  0.348694 -0.089532   

             V24       V25       V26       V27       V28  Amount  Class  
7014    0.111299  0.438692  0.517427 -0.010175  0.003499    0.01      0  
141758 -0.445667  0.594742 -0.632797  0.089174  0.025675    1.00      0  
236261 -0.120246 -0.446900 -0.936400  0.030065 -0.013667  100.00      0  
113513  1.053947 -0.155220 -0.060465 -0.003955 -0.072871   24.95      0  
255101  0.005502  0.433262 -0.500659 -0.030705 -0.044763  103.99      0  

[5 rows x 31 columns]

In [21]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [22]:
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64

In [23]:
new_dataset.groupby('Class').mean() ## check whether good or bad sample 

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94694.386179 -0.118660 -0.051047 -0.023232 -0.022033 -0.050595   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.047684  0.057408 -0.000819  0.072899  ... -0.014780 -0.003794   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.023094 -0.045472 -0.004881 -0.013181  0.054298 -0.002401 -0.011171   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       94.333923  
1      122.211321  

[2 rows x 30 columns]

In [1]:
## spliting the data into Features $ Targets

In [24]:
X = new_dataset.drop(columns = 'Class',axis=1)
Y = new_dataset['Class']

In [26]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
7014      9195.0  1.174829 -0.027781  1.028653  1.056980 -0.552270  0.189836   
141758   84482.0  1.305574 -0.408376  0.548300 -0.917218 -0.847102 -0.357029   
236261  148737.0  1.811565 -0.665829 -0.466436  0.422015 -0.900105 -0.818862   
113513   73095.0  0.077719 -0.163961 -0.073190 -2.069075  2.015091  3.419668   
255101  157064.0  1.844602 -0.051082 -1.690806  1.465939  0.359141 -1.108321   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [27]:
print(Y)

7014      0
141758    0
236261    0
113513    0
255101    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


In [28]:
# spliting the feture and target data into Training and Test data 

In [29]:
X_train, X_test, Y_train, Y_test =train_test_split(X,Y,test_size = 0.2, stratify = Y, random_state =2)

In [30]:
print(X.shape,X_train.shape,X_test.shape)

(984, 30) (787, 30) (197, 30)


In [32]:
#Model training 
# logistic Regression 

In [41]:
model = LogisticRegression(solver='liblinear', max_iter=1000)

In [42]:
# training our logisticRegresstion model with training Data 
model.fit(X_train,Y_train)

LogisticRegression(max_iter=1000, solver='liblinear')

In [43]:
## Model Evaluation based on accuracy score 

In [44]:
X_train_prediction = model.predict(X_train) ## find a target value of xtrain data 
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)     ## find a accuracy for ture tatget and find targrt 

In [46]:
print('Accuracy on Training Data: ',training_data_accuracy ) 

Accuracy on Training Data:  0.9339263024142312


In [47]:
# accuracy of test data 

In [48]:
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [49]:
print('Accuracy on Test Data: ',test_data_accuracy ) 

Accuracy on Test Data:  0.9187817258883249


In [50]:
## F1

In [52]:
print(classification_report(Y_test, X_test_prediction))

              precision    recall  f1-score   support

           0       0.87      0.98      0.92        99
           1       0.98      0.86      0.91        98

    accuracy                           0.92       197
   macro avg       0.93      0.92      0.92       197
weighted avg       0.93      0.92      0.92       197



#Insights & Analysis
1. High Overall Accuracy:
Model performs well on both training and test sets, indicating it generalizes well and is not overfitting.
Test accuracy of 91.88% is a solid performance, especially for an imbalanced binary classification task.

2. Balanced Performance Across Classes:
Class 0 (Legit): Precision = 0.87, Recall = 0.98 → most normal transactions are correctly identified.
Class 1 (Fraud): Precision = 0.98, Recall = 0.86 → the model is great at correctly predicting fraud cases and has a high precision, meaning fewer false alarms.

3. F1-Score Indicates Strong Balance:
The F1-score for fraud detection (0.91) shows that the model maintains a strong balance between catching frauds and minimizing false positives.
The macro and weighted averages are close (both ~0.92), suggesting consistent model behavior across both classes.
